lib/connections/structs.go

```diff
@@ -88,15 +92,29 @@ func (t connType) Transport() string {
 }
 
 func newInternalConn(tc tlsConn, connType connType, isLocal bool, priority int) internalConn {
+	now := time.Now()
 	return internalConn{
 		tlsConn:       tc,
 		connType:      connType,
 		isLocal:       isLocal,
 		priority:      priority,
-		establishedAt: time.Now().Truncate(time.Second),
+		establishedAt: now.Truncate(time.Second),
+		connectionID:  newConnectionID(tc, connType, now),
 	}
 }
 
+// newConnectionID generates a connection ID. The connection ID is designed
+// to be 1) unique for each connection (even those reusing the same socket
+// address on both sides), 2) sortable so that the connection with the
+// lowest ID will be the primary one. This also coincides with being the
+// oldest connection.
```

Comment:
In principle there's no strict requirement that the same connection is considered primary on both sides (but it's better, so we avoid delaying index messages with request/responses), we just need to ensure that a given side considers *one* connection primary so that index updates are properly serialized. In practice, we assume that index updates will come in on the primary connection, but if we get index updates somewhere else we move the index handler there (in ensureIndexHandler) so things work out over time. (We should probably mark that connection as promoted then as well, which I see I don't, I'll look into that.)

Reply:
So, this has changed, and the way it works now is that both sides send a timestamp, and then each side builds a connection ID based on the sum of the timestamps. This way, both sides gets the same number and will sort the connections the same. It's also somewhat resilient to one side fucking it up (for example always sending `42` as the timestamp), although if the other side is malicious (send `-time.Now().UnixNano()` for example) things will end up less sorted. That said it should work mostly anyway as each side tacks on a random part to avoid the ID being entirely controllable by the other side. (The one thing we need to avoid is collisions, we don't want two connections with the same ID.)